kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionBuilderImpl.java

```diff
@@ -188,25 +189,145 @@ public Transaction build(Engine engine) {
       if (operation == Operation.CREATE_TABLE) {
         throw new TableAlreadyExistsException(table.getPath(engine), "Operation = CREATE_TABLE");
       }
+      return buildTransactionInternal(engine, false /* isNewTableDef */, Optional.of(snapshot));
     } catch (TableNotFoundException tblf) {
       String tablePath = table.getPath(engine);
       logger.info("Table {} doesn't exist yet. Trying to create a new table.", tablePath);
       schema.orElseThrow(() -> requiresSchemaForNewTable(tablePath));
-      // Table doesn't exist yet. Create an initial snapshot with the new schema.
-      Metadata metadata = getInitialMetadata();
-      Protocol protocol = getInitialProtocol();
-      SnapshotQueryContext snapshotContext = SnapshotQueryContext.forVersionSnapshot(tablePath, -1);
-      LogReplay logReplay =
-          getEmptyLogReplay(engine, metadata, protocol, snapshotContext.getSnapshotMetrics());
-      snapshot =
-          new InitialSnapshot(table.getDataPath(), logReplay, metadata, protocol, snapshotContext);
+      return buildTransactionInternal(engine, true /* isNewTableDef */, Optional.empty());
     }
+  }
+
+  /**
+   * Returns a built {@link Transaction} for this transaction builder (with the input provided by
+   * the user) given the provided parameters. This includes validation and updates as defined in the
+   * builder.
+   *
+   * @param isNewTableDef whether we are defining a new table definition or not. This determines
+   *     what metadata to commit in the returned transaction, and what operations to allow or block.
+   * @param existingSnapshot the existing latest snapshot of the table (or empty for a new table).
+   *     This is used to validate that we can write to the table, and to get the protocol/metadata
+   *     when isNewTableDef=false.
+   */
+  protected TransactionImpl buildTransactionInternal(
+      Engine engine, boolean isNewTableDef, Optional<SnapshotImpl> existingSnapshot) {
+    checkArgument(
+        isNewTableDef || existingSnapshot.isPresent(),
+        "Existing snapshot must be provided if not defining a new table definition");
+    Metadata baseMetadata;
+    Protocol baseProtocol;
+    if (isNewTableDef) {
+      // For a new table definition start with an empty initial metadata
+      baseMetadata = getInitialMetadata();
+    } else {
+      // Otherwise, use the existing table metadata
+      baseMetadata = existingSnapshot.get().getMetadata();
+    }
+    if (existingSnapshot.isPresent()) {
+      // If the table exists, start with the prior table protocol - this ensures we never downgrade
+      baseProtocol = existingSnapshot.get().getProtocol();
+    } else {
+      // Otherwise, start with initial protocol
+      baseProtocol = getInitialProtocol();
+    }
```

Comment:
These two if statements do basically the same thing but they are checking different values for the `if` -- can we make these check the same value? i.e. `if (existingSnapshot.isPresent()) {` seems good to me for both

Reply:
We actually have different behavior here; for replace we want to start with initial empty metadata but reuse the protocol. For create we want initial empty for both